### PR TITLE
[BugFix] Error in documentation's numpy differentiation code example

### PR DESCRIPTION
**Pull request by mariaschuld**
**Context:**

Updates a failing code example.

**Description of the Change:**

Use keyword argument to mark a non-differentiable wires argument.
Add a random seed to make example reproducable.

Updated example reads:

Screenshot from 2021-08-05 12-34-00

**Related GitHub Issues:**
#1498

### CODE REVIEW

File: doc/introduction/interfaces/numpy.rst [hunk 1]
```diff
@@ -163,20 +163,22 @@ and two non-trainable arguments ``data`` and ``wires``:
         qml.CNOT(wires=[wires[0], wires[2]])
         return qml.expval(qml.PauliZ(wires[0]))
 
-We must specify that ``data`` and ``wires`` are NumPy arrays with ``requires_grad=False``:
+Since ``wires`` is a keyword argument, it will be automatically non-trainable. However, we
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:166]: @josh146 will this also be true after we give up the backwards compatibility hack?

File: doc/introduction/interfaces/numpy.rst [hunk 2]
```diff
@@ -163,20 +163,22 @@ and two non-trainable arguments ``data`` and ``wires``:
...
+must make ``data`` non-trainable by specifying it as a NumPy array with ``requires_grad=False``:
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:167]:
```suggestion
must explicitly make ``data`` non-trainable by specifying it as a NumPy array with ``requires_grad=False``:
```

File: doc/introduction/interfaces/numpy.rst [hunk 3]
```diff
@@ -163,20 +163,22 @@ and two non-trainable arguments ``data`` and ``wires``:
...
->>> circuit(weights, data, wires)
-0.16935626052294817
+>>> wires = [2, 0, 1]
+>>> circuit(weights, data, wires=wires)
+0.4124409353413991
 
-When we compute the derivative, arguments with ``requires_grad=False`` are explicitly ignored
-by :func:`~.grad`:
+When we compute the derivative, arguments with ``requires_grad=False``, as well as keyword arguments,
+are explicitly ignored by :func:`~.grad`:
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:177]:
```suggestion
are ignored by :func:`~.grad`:
```

File: doc/introduction/interfaces/numpy.rst [hunk 4]
```diff
@@ -154,7 +154,7 @@ and two non-trainable arguments ``data`` and ``wires``:
     dev = qml.device('default.qubit', wires=5)
 
     @qml.qnode(dev)
-    def circuit(weights, data, wires):
+    def circuit(weights, data, wires=None):
```
Comment (josh146) [inline doc/introduction/interfaces/numpy.rst:157]:
```suggestion
    def circuit(weights, data, wires):
```

This can stay as is --- the workaround is just that it needs to be _passed_ as a kwarg!

File: doc/introduction/interfaces/numpy.rst (hunk 1 above)
Comment (josh146) [reply doc/introduction/interfaces/numpy.rst:166]: Nope! In fact, this isn't even a PL thing, but an autograd thing - it is autograd that is ignoring keyword arguments 😆 So maybe best to reword this?

File: doc/introduction/interfaces/numpy.rst [hunk 5]
```diff
@@ -163,20 +163,22 @@ and two non-trainable arguments ``data`` and ``wires``:
...
+must explicitly make ``data`` non-trainable by specifying it as a NumPy array with ``requires_grad=False``:
 
+>>> np.random.seed(42)  # make the results reproducable
 >>> weights = np.array([0.1, 0.2, 0.3])
 >>> data = np.random.random([2**3], requires_grad=False)
->>> wires = np.array([2, 0, 1], requires_grad=False)
->>> circuit(weights, data, wires)
-0.16935626052294817
+>>> wires = [2, 0, 1]
+>>> circuit(weights, data, wires=wires)
```
Comment (josh146) [inline doc/introduction/interfaces/numpy.rst:173]: As far as I am aware, this is the only code change necessary to make it work, the change above (`wires=None`) should not be needed

File: doc/introduction/interfaces/numpy.rst [hunk 6]
```diff
@@ -163,20 +163,22 @@ and two non-trainable arguments ``data`` and ``wires``:
...
+0.4124409353413991
 
-When we compute the derivative, arguments with ``requires_grad=False`` are explicitly ignored
-by :func:`~.grad`:
+When we compute the derivative, arguments with ``requires_grad=False``, as well as keyword arguments,
+are ignored by :func:`~.grad`:
```
Comment (josh146) [inline doc/introduction/interfaces/numpy.rst:177]: 👍 

I think this is all that needs to be said?

**Review by josh146 — CHANGES_REQUESTED** [review]
Thanks for getting to this so quickly @mariaschuld! I think we also need to solve the underlying issue (the backwards compatibility 'everything in a positional argument is trainable') ASAP

File: doc/introduction/interfaces/numpy.rst [hunk 7]
```diff
@@ -122,12 +122,12 @@ Differentiable and non-differentiable arguments
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
 How does PennyLane know which arguments of a quantum function to differentiate, and which to ignore?
-For example, you may want to pass arguments as positional arguments to a QNode but *not* have
+For example, you may want to pass arguments to a QNode but *not* have
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:125]: I decided to polish the entire paragraph, so it makes sense together. I think here `as positional arguments` is confusing because we only state the rule about positional arguments below.

File: doc/introduction/interfaces/numpy.rst [hunk 8]
```diff
@@ -122,12 +122,12 @@ Differentiable and non-differentiable arguments
...
 PennyLane consider them when computing gradients.
 
 As a basic rule, **all positional arguments provided to the QNode are assumed to be differentiable
-by default**. To accomplish this, all arrays created by the PennyLane NumPy module have a special
-flag ``requires_grad`` specifying whether they are trainable or not:
+by default**. To accomplish this, arguments are internally turned into arrays of the PennyLane NumPy module,
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:129]: I think this piece of information is super important!

File: doc/introduction/interfaces/numpy.rst [hunk 9]
```diff
@@ -146,8 +146,11 @@ tensor([0.1, 0.2], requires_grad=False)
     The ``requires_grad`` argument can be passed to any NumPy function provided by PennyLane,
     including NumPy functions that create arrays like ``np.random.random``, ``np.zeros``, etc.
 
+An alternative way to avoid differentiation logic to be applied to positional arguments is to
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:149]:
```suggestion
An alternative way to avoid having positional arguments turned into differentiable PennyLane NumPy arrays is to
```

File: doc/introduction/interfaces/numpy.rst [hunk 10]
```diff
@@ -163,20 +166,28 @@ and two non-trainable arguments ``data`` and ``wires``:
...
+
+>>> np.random.seed(42)  # make the results reproducable
 >>> data = np.random.random([2**3], requires_grad=False)
->>> wires = np.array([2, 0, 1], requires_grad=False)
->>> circuit(weights, data, wires)
-0.16935626052294817
 
-When we compute the derivative, arguments with ``requires_grad=False`` are explicitly ignored
-by :func:`~.grad`:
+But ``wires`` is a list in this example, and if we turned it into a PennyLane NumPy array we would have to
```
Comment (mariaschuld) [inline doc/introduction/interfaces/numpy.rst:175]:
```suggestion
But ``wires`` is a list in this example, and if we turn it into a PennyLane NumPy array we would have to
```

**Comment by mariaschuld** [comment]
@josh146 I updated the entire section to be clearer. We will only need to make one or two text adjustments when the upcoming change happens, but the examples should work either way. 

Do you think this is better?
